### ISSUE TITLE
Trouble with Docker on Windows 10

### ISSUE THREAD

**Issue by Croq360** [wouterdebie/locast2tuner#29]
Followed the instructions in the README (changing $HOME to the %USERPROFILE% variable which is what it is in Windows).  I got it into Docker fine, but when I try to run it, I get the following:

```
config selected = /app/config/config.ini
thread 'main' panicked at 'Username required', src/config.rs:70:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Could not load selected file Syntax
```

Note that I edited the config.ini that's in my .locast2tuner folder.  Maybe it's not using that, based on the error message?  Perhaps there's a config.ini somewhere else I need to be editing.  Any ideas or tips?

**Comment by eiddor** [comment]
@Croq360 I suspect you'll have to Windowsify things a bit with Docker, starting with maybe not using a dot directory.

Are you using the `docker run` command or docker-compose?

It might also be a CR/LF issue.

Try these steps:

- create your own `config.ini` from scratch using Notepad, matching the syntax from the sample file
- put it in the current directory that you're running from
- modify the the volume mapping in your `docker run` command to be `-v .:/app/config`
- if you're using docker-compose, modify the volume section to look like this:

```yaml
    volumes:
      - .:/app/config
``` 

Paying attention to spacing, since it's YAML.

I have a headless Windows box that I'll fire-up and test with here in a few minutes.

**Comment by eiddor** [comment]
Oh goody, finally a reason to install WSL2.

**Comment by eiddor** [comment]
Yeah, it's a CR/LF/newline/whatever issue.  Grr.

Run `dos2unix` on your config file before you launch the container and you should be fine.

**Comment by Croq360** [comment]
Ok, will give that a try. Running the exe directly and pointing to a config file that's like what was on locast2dvr works fine.  I'll use my VS Code editor or notepad++ to save the config that I downloaded and edited with my info and put into .locast2tuner folder. Will report back on that.

**Comment by eiddor** [comment]
Yeah, that makes sense - The EXE is fine reading the ^Ms in the file put there by Windows, but the Docker version is technically running in Linux, so it chokes on them.

Either of those editors should be capable of saving in Unix format and stripping the CRs, so you should be ok.  That's all `dos2unix` does.

I added a note to the documentation.

**Comment by Croq360** [comment]
So I've tried checking the config.ini file, and it already has Unix end of lines (LF only).  I even tried saving with DOS (CR/LF) and then back to Unix (LF) format.  No good.

So, now I'm playing around with using WinSW to install it as a service.  Whenever I'm running there though, I'm getting the following error:
Child process 'locast2tuner (4368)' finished with code 101.

I need to look through the code here and figure out if 101 is a specific error, and see if that gives me a clue on how to fix my Service.

**Comment by Croq360** [comment]
So I changed the backslashes ("\") in the paths to forward slashes ("/") in the XML config for WinSW, and the service is running now.

**Comment by eiddor** [comment]
Wow, interesting - I had the exact issue as you did with "Could not load file syntax" and removing the CRs got rid of it.  I did have to massage my `docker-compose.yml` file to adjust the volume location, too.

Glad you got it working!  Might be worth documenting in `README.md` or for @wouterdebie to use for the Windows package.

**Comment by Croq360** [comment]
Yeah, once the service runs for a while and I confirm everything is working okay, I'll share my configuration for the service along with notes/comments.

**Comment by wouterdebie** [comment]
Great work! 
Just for my understanding, you are running a Linux Docker image on Windows?

In the case of a windows executable, I'd like to accompany that with making `locast2tuner` run as a service.

**Comment by Croq360** [comment]
In my case, I was trying to use the Linux Docker image on Windows (which it should be able to do).  I'm pretty sure when I did that with locast2plex, it ran in a Linux container.

Regarding the Windows executable/service, for Windows users, I can put together some documentation on getting locast2tuner executable running as a service using WinSW (https://github.com/winsw/winsw).  I'm using it instead of NSSM since I had some issues with NSSM and locast2dvr locking up after a while and being totally unresponsive after a time.  I should know in a few days if it looks like WinSW will be the option to go with.

**Comment by Croq360** [comment]
I think I see the problem.  When I ran the Windows equivalent for the curl command, "curl -o %USERPROFILE%\.locast2tuner\config.ini https://github.com/wouterdebie/locast2tuner/blob/main/assets/config.ini.example", it tacked on a TON of extra HTML tagging.  When I look at the example file here, it just has the entries called out.  When I look at my resulting config.ini file, it's nearly 1200 lines long, with the "config" stuff put into an html table like this:

```html
<table class="highlight tab-size js-file-line-container" data-tab-size="8" data-paste-markdown-skip>
      <tr>
        <td id="L1" class="blob-num js-line-number" data-line-number="1"></td>
        <td id="LC1" class="blob-code blob-code-inner js-file-line"># Mandatory configuration</td>
      </tr>
      <tr>
        <td id="L2" class="blob-num js-line-number" data-line-number="2"></td>
        <td id="LC2" class="blob-code blob-code-inner js-file-line">username=&quot;my_locast@username&quot;</td>
      </tr>
      <tr>
        <td id="L3" class="blob-num js-line-number" data-line-number="3"></td>
        <td id="LC3" class="blob-code blob-code-inner js-file-line">password=&quot;very_secret_password&quot;</td>
      </tr>
      <tr>
        <td id="L4" class="blob-num js-line-number" data-line-number="4"></td>
        <td id="LC4" class="blob-code blob-code-inner js-file-line">
</td>
      </tr>
      <tr>
        <td id="L5" class="blob-num js-line-number" data-line-number="5"></td>
        <td id="LC5" class="blob-code blob-code-inner js-file-line"># Optional configuration</td>
      </tr>
      <tr>
        <td id="L6" class="blob-num js-line-number" data-line-number="6"></td>
        <td id="LC6" class="blob-code blob-code-inner js-file-line"># bind_address = &quot;127.0.0.1&quot;</td>
      </tr>
      <tr>
        <td id="L7" class="blob-num js-line-number" data-line-number="7"></td>
        <td id="LC7" class="blob-code blob-code-inner js-file-line"># override_zipcodes=&quot;98101,94102,90210,85006,80123,57701,57103,51054,55111,75001,77002,60601,46077,30301,33602,33125,33403,20002,21206,19103,18505,10009,02115&quot;</td>
      </tr>
      <tr>
        <td id="L8" class="blob-num js-line-number" data-line-number="8"></td>
        <td id="LC8" class="blob-code blob-code-inner js-file-line"># days = 8</td>
      </tr>
      <tr>
        <td id="L9" class="blob-num js-line-number" data-line-number="9"></td>
        <td id="LC9" class="blob-code blob-code-inner js-file-line"># verbose = 1</td>
      </tr>
      <tr>
        <td id="L10" class="blob-num js-line-number" data-line-number="10"></td>
        <td id="LC10" class="blob-code blob-code-inner js-file-line"># port = 6077</td>
      </tr>
      <tr>
        <td id="L11" class="blob-num js-line-number" data-line-number="11"></td>
        <td id="LC11" class="blob-code blob-code-inner js-file-line"># tuner_count = 3</td>
      </tr>
      <tr>
        <td id="L12" class="blob-num js-line-number" data-line-number="12"></td>
        <td id="LC12" class="blob-code blob-code-inner js-file-line"># disable_station_cache = false</td>
      </tr>
      <tr>
        <td id="L13" class="blob-num js-line-number" data-line-number="13"></td>
        <td id="LC13" class="blob-code blob-code-inner js-file-line"># multiplex = false</td>
      </tr>
      <tr>
        <td id="L14" class="blob-num js-line-number" data-line-number="14"></td>
        <td id="LC14" class="blob-code blob-code-inner js-file-line"># remap = false</td>
      </tr>
      <tr>
        <td id="L15" class="blob-num js-line-number" data-line-number="15"></td>
        <td id="LC15" class="blob-code blob-code-inner js-file-line"># ssdp = false</td>
      </tr>
      <tr>
        <td id="L16" class="blob-num js-line-number" data-line-number="16"></td>
        <td id="LC16" class="blob-code blob-code-inner js-file-line"># cache_dir = &quot;/var/lib/locast2tuner&quot; # default: $HOME/.locast2tuner</td>
      </tr>
      <tr>
        <td id="L17" class="blob-num js-line-number" data-line-number="17"></td>
        <td id="LC17" class="blob-code blob-code-inner js-file-line"># cache_timeout = 3600</td>
      </tr>
      <tr>
        <td id="L18" class="blob-num js-line-number" data-line-number="18"></td>
        <td id="LC18" class="blob-code blob-code-inner js-file-line"># device_firmware = &quot;homerun3_atsc&quot;</td>
      </tr>
      <tr>
        <td id="L19" class="blob-num js-line-number" data-line-number="19"></td>
        <td id="LC19" class="blob-code blob-code-inner js-file-line"># device_model = &quot;HDHR3-US&quot;</td>
      </tr>
      <tr>
        <td id="L20" class="blob-num js-line-number" data-line-number="20"></td>
        <td id="LC20" class="blob-code blob-code-inner js-file-line"># device_version = &quot;20170612&quot;</td>
      </tr>
</table>
```

**Comment by wouterdebie** [comment]
My bad.. the link should have been to the raw file, which I just fixed in commit c7f1d0a.

**Comment by Croq360** [comment]
Awesome, thanks!
